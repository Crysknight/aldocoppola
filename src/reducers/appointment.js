import { CHOOSE_CENTER, CHOOSE_EMPLOYEE, CONFIRM_SERVICES, CONFIRM_DATE_TIME } from '../actions/types';

export default (state = {}, action) => {
	switch (action.type) {
		case CHOOSE_CENTER: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.centerChosen = action.payload;
			return newState;
		}
		case CHOOSE_EMPLOYEE: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.employeeChosen = action.payload;
			return newState;
		}
		case CONFIRM_SERVICES: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.servicesChosen = action.payload;
			return newState;
		}
		case CONFIRM_DATE_TIME: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.dateTimeChosen = action.payload;
			return newState;
		}
		default: {
			return state;
		}
	}
};