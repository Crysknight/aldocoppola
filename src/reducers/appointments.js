import { 
	ADD_APPOINTMENT, 
	CANCEL_NEW_APPOINTMENT, 
	CHOOSE_CENTER,
	DELETE_APPOINTMENT,
	CONFIRM_APPOINTMENT_EDIT
} from '../actions/types';

export default (state = [], action) => {
	switch (action.type) {
		case ADD_APPOINTMENT: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.push(action.payload);
			return newState;
		}
		case CANCEL_NEW_APPOINTMENT: {
			if (action.payload.isTheLastOneTheOnlyOne) {
				return [];
			} else {
				return state;
			}
		}
		case CHOOSE_CENTER: {
			return [];
		}
		case DELETE_APPOINTMENT: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let i = 0; i < newState.length; i++) {
				if (newState[i].number === action.payload.appointment.number) {
					newState.splice(i, 1);
				}
			}
			for (let i = 0; i < newState.length; i++) {
				newState[i].number = i;
			}
			return newState;
		}
		case CONFIRM_APPOINTMENT_EDIT: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let i = 0; i < newState.length; i++) {
				if (newState[i].number === action.payload.appointment.number) {
					delete action.payload.appointment.edited;
					newState[i] = action.payload.appointment;
				}
			}
			return newState;
		}
		default: {
			return state;
		}
	}
}