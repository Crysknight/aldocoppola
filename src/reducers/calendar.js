import moment from 'moment';
import 'moment/locale/ru';

import { CHOOSE_EMPLOYEE, CHOOSE_DATE, CHOOSE_TIME } from '../actions/types';

const getCalendar = (numberOfMonths) => {

	const calendar = [];

	for (let i = moment().month(); i < (moment().month() + numberOfMonths); i++) {

		let number = i % 12;
		let year = i < 12 ? moment().year() : moment().year() + 1;
		let name = moment([year, number]).format('MMMM');
		name = name.charAt(0).toUpperCase() + name.slice(1);
		let days = [];
		let numberOfDays = moment([year, number]).daysInMonth();
		for (let j = 1; j <= numberOfDays; j++) {
			let dayOfWeek = +moment([year, number, j]).format('d') - 1;
			let dayOfWeekName = moment([year, number, j]).format('dddd');
			dayOfWeekName = dayOfWeekName.charAt(0).toUpperCase() + dayOfWeekName.slice(1);
			let day = moment([year, number, j]);
			let today = moment([moment().year(), moment().month(), moment().date()]);
			days.push({
				date: j,
				dayOfWeek: dayOfWeek === -1 ? 6 : dayOfWeek,
				dayOfWeekName,
				past: moment(today).isAfter(day)
			});
		}
		calendar.push({
			number,
			name,
			year,
			days
		});

	}

	return calendar;

}

window.moment = moment;

export default (state = getCalendar(8), action) => {
	switch (action.type) {
		case CHOOSE_EMPLOYEE: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let workHour of action.payload.workHoursFree) {
				workHour = moment(workHour, 'YYYY-MM-DD hh:mm');
				let month = workHour.month();
				let date = workHour.date();
				for (let calendarMonth of newState) {
					if (calendarMonth.number === month) {
						for (let calendarDate of calendarMonth.days) {
							if (calendarDate.date === date) {
								calendarDate.workHours = calendarDate.workHours ? calendarDate.workHours : {};
								let morningBorder = workHour.clone().hour(12).minute(0);
								let dayBorder = workHour.clone().hour(18).minute(0);
								if (workHour.isSameOrBefore(morningBorder)) {
									calendarDate.workHours["Утро"] = calendarDate.workHours["Утро"] ? calendarDate.workHours["Утро"] : [];
									calendarDate.workHours["Утро"].push(workHour.format('HH:mm'));
								} else if (workHour.isSameOrBefore(dayBorder)) {
									calendarDate.workHours["День"] = calendarDate.workHours["День"] ? calendarDate.workHours["День"] : [];
									calendarDate.workHours["День"].push(workHour.format('HH:mm'));
								} else {
									calendarDate.workHours["Вечер"] = calendarDate.workHours["Вечер"] ? calendarDate.workHours["Вечер"] : [];
									calendarDate.workHours["Вечер"].push(workHour.format('HH:mm'));
								}
								// calendarDate.workHours.push(workHour.format('HH:mm'));
							}
						}
					}
				}
			}
			return newState;
		}
		case CHOOSE_DATE: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			let month = action.payload.month;
			let date = action.payload.date;
			for (let calendarMonth of newState) {
				calendarMonth.chosen = calendarMonth.chosen ? false : calendarMonth.chosen;
				for (let calendarDate of calendarMonth.days) {
					calendarDate.chosen = calendarDate.chosen ? false : calendarDate.chosen;
					calendarDate.timeChosen = calendarDate.timeChosen ? null : calendarDate.timeChosen;
				}
				if (calendarMonth.number === month) {
					calendarMonth.chosen = true;
					for (let calendarDate of calendarMonth.days) {
						if (calendarDate.date === date) {
							calendarDate.chosen = true;
						}
					}
				}
			}
			return newState;
		}
		case CHOOSE_TIME: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			let month = action.payload.date.monthNumber;
			let date = action.payload.date.date;
			for (let calendarMonth of newState) {
				calendarMonth.chosen = calendarMonth.chosen ? false : calendarMonth.chosen;
				for (let calendarDate of calendarMonth.days) {
					calendarDate.chosen = calendarDate.chosen ? false : calendarDate.chosen;
					calendarDate.timeChosen = calendarDate.timeChosen ? null : calendarDate.timeChosen;
				}
				if (calendarMonth.number === month) {
					calendarMonth.chosen = true;
					for (let calendarDate of calendarMonth.days) {
						if (calendarDate.date === date) {
							calendarDate.chosen = true;
							calendarDate.timeChosen = action.payload.time;
						}
					}
				}
			}
			return newState;
		}
		default: {
			return state;
		}
	}
};