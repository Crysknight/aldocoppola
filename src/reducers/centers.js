import { GOT_CENTERS } from '../actions/types';

export default (state = [], action) => {
	switch (action.type) {
		case GOT_CENTERS: {
			return action.payload;
		}
		default: {
			return state;
		}
	}
};