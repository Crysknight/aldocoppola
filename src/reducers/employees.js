import { GOT_EMPLOYEES, CHOOSE_CENTER } from '../actions/types';

const employeesMethods = {};

employeesMethods.findEmployeeById = (employees, id) => {
	for (let employee of employees) {
		if (employee.id === id) {
			return employee;
		}
	}
};

export { employeesMethods };

export default (state = [], action) => {
	switch (action.type) {
		case GOT_EMPLOYEES: {
			return action.payload;
		}
		case CHOOSE_CENTER: {
			return [];
		}
		default: {
			return state;
		}
	}
};