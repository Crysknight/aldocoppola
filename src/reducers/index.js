import { combineReducers } from 'redux';
import defaultReducer from './default';

const allReducers = combineReducers({
	defaultReducer: defaultReducer
});

export default allReducers;
