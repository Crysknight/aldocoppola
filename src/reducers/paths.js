import { GOT_EMPLOYEES, GOT_MAIN_SERVICES } from '../actions/types';

// Basis of all widget paths (after this everything is widget)
const appPath = '/_appliance/';

// Regular expression to find any path related to the widget
const appRegExp = new RegExp(appPath.slice(0, -1) + '(.*)');

// Object with paths used by the app
const paths = {
	__app: {
		pathString: '',
		privacy: true
	},
	ChooseCenter: {
		pathString: 'choose-center/',
		privacy: false
	},
	ChooseEmployee: {
		pathString: 'choose-employee/',
		privacy: true
	},
	ChooseServices: {
		pathString: 'choose-services/',
		privacy: true,
		childPaths: {
			SubmitServices: {
				pathString: 'submit-services/',
				privacy: true
			}
		}
	},
	ChooseDateTime: {
		pathString: 'choose-datetime/',
		privacy: true
	}
};

const pathsMethods = {};

// Function for constructing pathstrings from the sctructure above
// It doesn't check the 'pathString' consistency so please regard the following format
// pathString: 'nameofthepathstring/' with slash at the end
pathsMethods.createPathStrings = (paths, nested) => {
	// Firstly, it loops through the paths object, finding properties that are paths too
	for (let path in paths) {
		// Do not process the root path of the widget
		if (path === '__app') {
			paths[path].pathString = appPath;
			continue;
		}
		let parentString;
		// The 'finished' property denotes that we already processed this pathstring
		// not to process it in further calls of function
		if (nested || paths[path].finished) {
			parentString = paths[path].pathString;
		} else {
			// Prepending 'appPath' at first level of structure
			parentString = paths[path].pathString = appPath + paths[path].pathString;
			paths[path].finished = true;
		}
		// If the path has childPaths, we process every childPath's pathString
		if (paths[path].childPaths) {
			let childPaths = paths[path].childPaths;
			for (let path in childPaths) {
				if (!childPaths[path].finished) {
					childPaths[path].pathString = parentString + childPaths[path].pathString;
					childPaths[path].finished = true;
				}
				// And recursively process the childPaths too
				pathsMethods.createPathStrings(childPaths, true);
			}
		}
	}
};

pathsMethods.createPathStrings(paths);

// A function to get the path of the specific page
pathsMethods.getPath = (paths, currentPath, path) => {
	if (!path) {
		return currentPath.replace(appRegExp, '') + paths.__app.pathString;
	}
	return currentPath.replace(appRegExp, '') + path.pathString;
};

// A function to create a Link that opens and closes the app
pathsMethods.getAppSwitch = (paths, currentPath) => {
	if (currentPath.match(appRegExp)) {
		return currentPath.replace(appRegExp, '');
	} else {
		return currentPath.replace(/\/$/, '') + paths.__app.pathString;
	}
};

// A function to create dynamic paths. 
// Takes in a path object: properties are - 'pathName', 'pathString', 'parentPath' and 'privacy'
pathsMethods.createPath = (paths, path) => {
	if (!path.pathString || !path.pathName) {
		throw new Error('pathString, or pathName properties missing');
	}
	if (path.pathString.match(/ /) !== null) {
		throw new Error('pathString should not contain spaces');
	}
	if (path.pathString.match(/^(\w|\d|-|_)*\/$/) === null) {
		throw new Error('pathString should begin from a letter, number, \'-\', or \'_\' and end with one slash symbol');
	}
	let targetPath = { 
		pathString: path.pathString,
		privacy: path.hasOwnProperty('privacy') ? path.privacy : true
	}
	// 'hasOwnProperty' because the 'parentPath' itself can be undefined
	if (!path.hasOwnProperty('parentPath')) {
		paths[path.pathName] = targetPath;
	} else if (path.hasOwnProperty('parentPath') && !path.parentPath) {
		throw new Error('Specified parentPath not found');
	} else {
		// createPath's recursive function to search a parentPath inside the paths object
		// In every path that has childPaths it also searches the parentPath in them recursively calling itself 
		let pathsCrawler = function(paths, targetPath, pathName, parentPath) {
			for (let path in paths) {
				if (paths[path].pathString === parentPath.pathString) {
					paths[path].childPaths = paths[path].childPaths ? paths[path].childPaths : {};
					paths[path].childPaths[pathName] = targetPath;
				} else if (paths[path].childPaths) {
					pathsCrawler(paths[path].childPaths, targetPath, pathName, parentPath);
				}
			}
		}
		pathsCrawler(paths, targetPath, path.pathName, path.parentPath);
	}
	pathsMethods.createPathStrings(paths);
};

// Export the paths and pathsMethods objects for testing purposes
export { pathsMethods };
export { paths };
// window.pathsMethods = pathsMethods;
// window.paths = paths;
// console.log(paths);

// Reducer, providing the paths object to the whole app
export default (state = paths, action) => {
	switch (action.type) {
		case GOT_EMPLOYEES: {
			let payload = action.payload;
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let employee of payload) {
				let pathRecourse = 'employee' + employee.id;
				pathsMethods.createPath(newState, {
					pathName: pathRecourse,
					pathString: pathRecourse + '/',
					privacy: true,
					parentPath: newState.ChooseEmployee
				});
			}
			return newState;
		}
		case GOT_MAIN_SERVICES: {
			let payload = action.payload;
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let service of payload) {
				let pathRecourse = 'service' + service.id;
				pathsMethods.createPath(newState, {
					pathName: pathRecourse,
					pathString: pathRecourse + '/',
					privacy: true,
					parentPath: newState.ChooseServices
				});
			}
			return newState;
		}
		default: {
			return state;
		}
	}
};