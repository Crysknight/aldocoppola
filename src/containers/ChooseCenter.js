import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import actions from '../actions';

import Header from '../components/header';
import Center from '../components/center';
import Footer from '../components/footer';

import { pathsMethods } from '../reducers/paths';

class ChooseCenter extends Component {

	// constructor(props) {
	// 	super(props);
	// }

	componentWillMount() {
		this.props.loadCenters();
	}

	chooseCenter(id, name) {
		let paths = this.props.paths;
		this.props.chooseCenter(id, name);
		this.props.history.push(pathsMethods.getPath(paths, this.props.location.pathname, paths.__app));
	}

	renderCenters() {
		return this.props.centers.map((center, index) => {
			return (
				<Center
					key={index}
					center={center}
					chooseCenter={() => this.chooseCenter(center.id, center.name)}
				/>
			);
		});
	}

	render() {
		let paths = this.props.paths;
		return (
			<div id="choose_center">
				<Header title="Выбрать центр красоты" />
				{this.renderCenters()}
				<Footer className="coal">
					<Link 
						to={pathsMethods.getPath(paths, this.props.match.path, paths.MyAccount)}
						className="footer-link my-account-link"
					>Личный кабинет</Link>
				</Footer>
			</div>
		);
	}

}

function mapStateToProps(state) {
	return {
		paths: state.paths,
		centers: state.centers
	};
}

function matchDispatchToProps(dispatch) {
	return bindActionCreators({
		loadCenters: actions.loadCenters,
		chooseCenter: actions.chooseCenter
	}, dispatch);
}

export default connect(mapStateToProps, matchDispatchToProps)(ChooseCenter);