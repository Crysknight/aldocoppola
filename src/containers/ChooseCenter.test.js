import React from 'react';
import ChooseCenter from './ChooseCenter';
import renderer from 'react-test-renderer';
import { MemoryRouter, Route } from 'react-router-dom';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import allReducers from '../reducers';
import { paths } from '../reducers/paths';



describe('ChooseCenter', () => {


	it('renders component properly', () => {

		const store = createStore(allReducers, applyMiddleware(thunk));
		const component = renderer.create(
			<Provider store={store}>
				<MemoryRouter initialEntries={[ '/test/bauble.html' + paths.ChooseCenter.pathString ]}>
					<Route path={'/test/bauble.html' + paths.ChooseCenter.pathString} component={ChooseCenter} />
				</MemoryRouter>
			</Provider>
		);
		let tree = component.toJSON();

		expect(tree).toMatchSnapshot();
		
	});


});