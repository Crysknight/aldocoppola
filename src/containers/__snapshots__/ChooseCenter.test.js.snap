// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseCenter renders component properly 1`] = `
<div
  id="choose_center"
>
  <div
    className="ac-header"
  >
    <h3>
      Выбрать центр красоты
    </h3>
  </div>
  <div
    className="ac-content"
  />
  <div
    className="ac-footer coal"
  >
    <a
      className="footer-link my-account-link"
      href="/test/bauble.html/_appliance/"
      onClick={[Function]}
    >
      <svg
        id="svg_lock"
        viewBox="584 630 10 15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Group 24"
          transform="translate(584 630)"
        >
          <path
            className="cls-1"
            d="M7.5,18.375h-5a2.507,2.507,0,0,1-2.5-2.5V11.5A2.507,2.507,0,0,1,2.5,9h5A2.507,2.507,0,0,1,10,11.5v4.375A2.507,2.507,0,0,1,7.5,18.375Zm-5-8.125A1.254,1.254,0,0,0,1.25,11.5v4.375a1.254,1.254,0,0,0,1.25,1.25h5a1.254,1.254,0,0,0,1.25-1.25V11.5A1.254,1.254,0,0,0,7.5,10.25Z"
            data-name="Path 26"
            id="Path_26"
            transform="translate(0 -3.375)"
          />
          <path
            className="cls-1"
            d="M8.875,6.875H2.625A.59.59,0,0,1,2,6.25V3.75A3.761,3.761,0,0,1,5.75,0,3.721,3.721,0,0,1,9.5,3.75v2.5A.626.626,0,0,1,8.875,6.875ZM3.25,5.625h5V3.75a2.5,2.5,0,0,0-5,0Z"
            data-name="Path 27"
            id="Path_27"
            transform="translate(-0.75)"
          />
          <path
            className="cls-1"
            d="M7.25,15.5A1.25,1.25,0,1,1,8.5,14.25,1.254,1.254,0,0,1,7.25,15.5Zm0-1.312Z"
            data-name="Path 28"
            id="Path_28"
            transform="translate(-2.25 -4.875)"
          />
          <path
            className="cls-1"
            d="M7.625,18.125A.59.59,0,0,1,7,17.5V15.625a.625.625,0,0,1,1.25,0V17.5A.626.626,0,0,1,7.625,18.125Z"
            data-name="Path 29"
            id="Path_29"
            transform="translate(-2.625 -5.625)"
          />
        </g>
      </svg>
      Личный кабинет
    </a>
  </div>
</div>
`;
